x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestDeleteExpiredDataAction.java

```diff
@@ -18,14 +19,17 @@
 
 import static org.elasticsearch.rest.RestRequest.Method.DELETE;
 import static org.elasticsearch.xpack.ml.MachineLearning.BASE_PATH;
+import static org.elasticsearch.xpack.ml.MachineLearning.PRE_V7_BASE_PATH;
 
 public class RestDeleteExpiredDataAction extends BaseRestHandler {
 
     @Override
     public List<Route> routes() {
         return List.of(
-            new Route(DELETE, BASE_PATH + "_delete_expired_data/{" + Job.ID + "}"),
-            new Route(DELETE, BASE_PATH + "_delete_expired_data")
+            Route.builder(DELETE, BASE_PATH + "_delete_expired_data/{" + Job.ID + "}")
+                .replaces(DELETE, PRE_V7_BASE_PATH + "_delete_expired_data/{" + Job.ID + "}", RestApiVersion.V_7).build(),
```

Comment:
What are you referring to @benwtrent ? The `delete_expired_data` action has existed as a REST api earlier than 7.x.

Reply:
I can see what it is.  Delete expired data can be called with or without a job ID.  The option to call it with a job ID was added in 7.x and never had an `_xpack` fallback.  So this is adding BWC in 8.x that didn't exist in 7.x.

_However_, the problem with enforcing the distinction is that it will cause considerable complication for the tests.  At the moment the `yamlRestCompatTest` setup in 8.x just does a global replace of `ml.delete_expired_data` with `xpack-ml.delete_expired_data`.  Enforcing the distinction would mean having more complicated test setup.  Complexity costs money and I don't think it's worth it.  If somebody does start using the newly added deprecated `_xpack` endpoint in 8.x instead of the correct one then it's likely because they're already using the `_xpack` everywhere else for ML, are using the REST compatibility layer, and know they have to do a sweep of their code at some point to switch to undeprecated endpoints.  So we haven't really caused them any extra work.  Anybody who already uses undeprecated endpoints is extremely unlikely to accidentally use a new deprecated `_xpack` endpoint.

So although pedantically one route should have the compatibility layer and one shouldn't, it's less work for us to just add them both and remove them both together.